app/models/token.rb

```diff
@@ -0,0 +1,69 @@
+# frozen_string_literal: true
+
+class Token
+  attr_reader :token_string
+
+  def initialize(token_string)
+    @token_string = token_string
+    validate_token
+  end
+
+  def to_s
+    @token_string
+  end
+
+  def payload
+    @payload ||= if @token_string
+                   pubkey = expected_key
+                   return if pubkey.blank?
+
+                   JWT.decode(@token_string, pubkey, true, algorithm: 'RS256')[0]
+                 end
+  rescue JWT::ExpiredSignature
+    raise error_klass('Validation error: token has expired')
+  end
+
+  def expected_key
+    decoded_token = JWT.decode(@token_string, nil, false, algorithm: 'RS256') # throwing a decode error
+    kid = decoded_token[1]['kid']
+    OIDC::KeyService.get_key(kid)
+  rescue JWT::DecodeError => e
+    raise error_klass("Validation error: #{e.message}")
+  end
+
+  def validate_token
+    raise error_klass('Validation error: no payload to validate') unless payload
+    raise error_klass('Validation error: issuer') unless valid_issuer?
+    raise error_klass('Validation error: audience') unless valid_audience?
+    raise error_klass('Validation error: token has expired') unless valid_expiration?
+  end
+
+  def valid_issuer?
+    payload['iss'] == Settings.oidc.issuer
+  end
+
+  def valid_audience?
+    payload['aud'] == Settings.oidc.audience
+  end
+
+  def valid_expiration?
+    payload['exp'] >= Time.current.utc.to_i
+  end
```

Comment:
We probably don't need to do this at all since the jwt library is already throwing an error if the token is expired

Reply:
I would prefer to keep this in to fail early